src/Geometry/coordinates.jl

```diff
@@ -43,32 +52,75 @@ end
-    XYZPoint{FT}(xyp.x, xyp.y, zp.z)
-product_coordinates(xyp::XYPoint{FT}, z::FT) where {FT} =
-    XYZPoint{FT}(xyp.x, xyp.y, z)
+product_coordinates(xyp::XYPoint, zp::ZPoint) =
+    XYZPoint(promote(xyp.x, xyp.y, zp.z)...)
 
 # TODO: get rid of these and refactor to consistent point types
 ZPoint(pt::Cartesian3Point{FT}) where {FT} = ZPoint{FT}(pt.x3)
-XYPoint(pt::Cartesian2DPoint{FT}) where {FT} = XYPoint{FT}(pt.x1, pt.x2)
-XYZPoint(pt::Cartesian3DPoint{FT}) where {FT} =
+XYPoint(pt::Cartesian12Point{FT}) where {FT} = XYPoint{FT}(pt.x1, pt.x2)
+XYZPoint(pt::Cartesian123Point{FT}) where {FT} =
     XYZPoint{FT}(pt.x1, pt.x2, pt.x3)
 
 Cartesian3Point(pt::ZPoint{FT}) where {FT} = Cartesian3Point{FT}(pt.z)
-Cartesian2DPoint(pt::XYPoint{FT}) where {FT} = Cartesian2DPoint{FT}(pt.x, pt.y)
-Cartesian3DPoint(pt::XYZPoint{FT}) where {FT} =
-    Cartesian3DPoint{FT}(pt.x, pt.y, pt.z)
-
-components(p::AbstractPoint) = SVector(ntuple(i -> getfield(p, i), nfields(p)))
+Cartesian12Point(pt::XYPoint{FT}) where {FT} = Cartesian12Point{FT}(pt.x, pt.y)
+Cartesian123Point(pt::XYZPoint{FT}) where {FT} =
+    Cartesian123Point{FT}(pt.x, pt.y, pt.z)
+
+ncomponents(p::AbstractPoint) = nfields(p)
+components(p::AbstractPoint{FT}, i::Symbol) where {FT} = getfield(p, i)::FT
+components(p::AbstractPoint{FT}, i::Integer) where {FT} = getfield(p, i)::FT
+components(p::AbstractPoint) =
+    SVector(ntuple(i -> components(p, i), ncomponents(p)))
+
+_coordinate_type(ptyp::Type{Abstract1DPoint}, ::Val{1}) = ptyp
+_coordinate(p::Abstract1DPoint, ::Val{1}) = p
+
+_coordinate_type(::Type{XYPoint{FT}}, ::Val{1}) where {FT} = XPoint{FT}
+_coordinate_type(::Type{XYPoint{FT}}, ::Val{2}) where {FT} = YPoint{FT}
+_coordinate(p::XYPoint, ::Val{1}) = XPoint(p.x)
+_coordinate(p::XYPoint, ::Val{2}) = YPoint(p.y)
+
+_coordinate_type(::Type{XZPoint{FT}}, ::Val{1}) where {FT} = XPoint{FT}
+_coordinate_type(::Type{XZPoint{FT}}, ::Val{2}) where {FT} = ZPoint{FT}
+_coordinate(p::XZPoint, ::Val{1}) = XPoint(p.x)
+_coordinate(p::XZPoint, ::Val{2}) = ZPoint(p.z)
+
+_coordinate_type(::Type{XYZPoint{FT}}, ::Val{1}) where {FT} = XPoint{FT}
+_coordinate_type(::Type{XYZPoint{FT}}, ::Val{2}) where {FT} = YPoint{FT}
+_coordinate_type(::Type{XYZPoint{FT}}, ::Val{3}) where {FT} = ZPoint{FT}
+_coordinate(p::XYZPoint, ::Val{1}) = XPoint(p.x)
+_coordinate(p::XYZPoint, ::Val{2}) = YPoint(p.y)
+_coordinate(p::XYZPoint, ::Val{3}) = ZPoint(p.z)
+
+coordinate_type(ptyp::Type{<:AbstractPoint}, ax::Int) =
+    _coordinate_type(ptyp, Val(ax))
+coordinate_type(ptyp::Type{<:AbstractPoint}, ax::Integer) =
+    _coordinate_type(ptyp, Int(ax))
+
+coordinate(pt::AbstractPoint, ax::Int) = _coordinate(pt, Val(ax))
+coordinate(pt::AbstractPoint, ax::Integer) = _coordinate(pt, Int(ax))
+
+# TODO: we need to rationalize point operations with vectors
+Base.:(-)(p1::T, p2::T) where {T <: AbstractPoint} =
+    unionalltype(T)((components(p1) - components(p2))...)
```

Comment:
What does this return?

Reply:
This returns a point that promotes correctly. I'll see if this is still needed as I was trying to get the local geometry to work correctly with CoordinateTypes but there needs to be a more incremental approach as there are a lot of baked in assumptions currently.

Ideally this should return a vector type.

Reply:
A vector type would make sense in a box, but wouldn't e.g. on a sphere.

Where is addition/subtraction of coordinates used? Is it just for ranges?